kong/plugins/ai-prompt-template/handler.lua

```diff
@@ -0,0 +1,116 @@
+local sub            = string.sub
+--
+
+_M.PRIORITY = 773
+_M.VERSION = kong_meta.version
+
+
+local log_entry_keys = {
+  REQUEST_BODY = "ai.payload.original_request",
+}
+
+local function bad_request(msg)
+  kong.log.debug(msg)
+  return kong.response.exit(ngx.HTTP_BAD_REQUEST, { error = { message = msg } })
+end
+
+local BRACE_START = byte("{")
+local BRACE_END = byte("}")
+local COLON = byte(":")
+local SLASH = byte("/")
+
+---- BORROWED FROM `kong.pdk.vault`
+---
+-- Checks if the passed in reference looks like a reference.
+-- Valid references start with '{template://' and end with '}'.
+--
+-- @local
+-- @function is_reference
+-- @tparam string reference reference to check
+-- @treturn boolean `true` is the passed in reference looks like a reference, otherwise `false`
+local function is_reference(reference)
+  return type(reference)      == "string"
+     and byte(reference, 1)   == BRACE_START
+     and byte(reference, -1)  == BRACE_END
+     and byte(reference, 10)   == COLON
+     and byte(reference, 11)   == SLASH
+     and byte(reference, 12)   == SLASH
+     and sub(reference, 2, 9) == "template"
+end
+
+local function find_template(reference_string, templates)
+  if is_reference(reference_string) then
+    local parts, err = parse_url(sub(reference_string, 2, -2))
+    if not parts then
+      return nil, fmt("template reference is not in format '{template://template_name}' (%s) [%s]", err, reference_string)
+    end
+
+    -- iterate templates to find it
+    for i, v in ipairs(templates) do
+      if v.name == parts.host then
+        return v, nil
+      end
+    end
+
+    return nil, "could not find template name [" .. parts.host .. "]"
+  end
+
+  return nil, "'messages' template reference should be a single string, in format '{template://template_name}'"
+end
+
+function _M:access(conf)
+  kong.service.request.enable_buffering()
+  kong.ctx.shared.ai_prompt_templated = true
+
+  if conf.log_original_request then
+    kong.log.set_serialize_value(log_entry_keys.REQUEST_BODY, kong.request.get_raw_body())
+  end
+
+  local request, err = kong.request.get_body("application/json")
+  if err then
+    return bad_request("this LLM route only supports application/json requests")
+  end
+
+  if (not request.messages) and (not request.prompt) then
+    return bad_request("this LLM route only supports llm/chat or llm/completions type requests")
+  end
+
+  if request.messages and request.prompt then
+    return bad_request("cannot run 'messages' and 'prompt' templates at the same time")
+  end
+
+  local reference
+  if request.messages then
+    reference = request.messages
+
+  elseif request.prompt then
+    reference = request.prompt
+
+  else
+    return bad_request("only 'llm/v1/chat' and 'llm/v1/completions' formats are supported for templating")
+  end
+
+  local requested_template, err = find_template(reference, conf.templates)
+  if err and (not conf.allow_untemplated_requests) then bad_request(err) end
```

Comment:
So you have a case where there are 3 results from `find_template` to consider: template exists, template doesn't exist, and some other semantic error. Currently you are not catching the case where there is a semantic error but `conf.allow_untemplated_requests` is enabled. The behavioral consequences of not handling this distinct case are probably quite trivial in most cases, but by not being explicit from the outset it means that fixing the behavior later on could _technically_ be considered a breaking change. Best to save future maintainers the headache.

There are a couple idiomatic ways to go about this:

#### alter `find_template` such that it returns `false` if the template is not found and `nil` for other errors


```lua
  local requested_template, err = find_template(reference, conf.templates)

  if requested_template == nil then
    return bad_request(err)

  elseif requested_template == false and not conf.allow_untemplated_requests then
    return bad_request(err)
  end
```

#### use a well-known error message for the "not found" case and match on it instead

```lua
  local requested_template, err = find_template(reference, conf.templates)

  if requested_template == nil then
    return bad_request(err)

  elseif err == NOT_FOUND and not conf.allow_untemplated_requests then
    return bad_request(err)
  end
```

Reply:
I've adjusted this a ton of times, and there isn't really a better way to write it other than like this.

Reply:
Ah no, so I've moved the "is_reference" out of the block, to make this easier to read now. See latest commit.